Добавлена блокировка пользователя администратором

NRenat opened the pull request:
- Добавлено скрытие профиля пользователя и профиля его коливинга в боте, если пользователь был заблокирован.
- В жалобы добавлены кнопки "Принять", "Отклонить", которые автоматически блокируют пользователя/меняют статус жалобы/скрывают его профиля в боте.
- Добавлено ограничение доступа к боту, если пользователь заблокирован.
- Изменен шаблон preview для лучшей читабельности.

NRenat committed:
Добавлена блокировка пользователя администратором
- Добавлено скрытие профиля пользователя и профиля его коливинга в боте, если пользователь был заблокирован.
- В жалобы добавлены кнопки "Принять", "Отклонить", которые автоматически блокируют пользователя, меняют статус жалобы и скрывают его профиля в боте.
- Добавлено ограничение доступа к боту, если пользователь заблокирован.
- Изменен шаблон preview для лучшей читабельности.

NRenat committed:
Удален метод .format()

NRenat committed:
Исправлена ошибка: новый пользователь не мог зарегистрироваться.

TomatoInOil commented on src/bot/conversations/profile/callback_funcs.py:
```diff
@@ -24,10 +26,15 @@ async def start(
+        banned_user = await check_user_ban(update)
+        if banned_user:
+            await update.callback_query.answer(text=BANNED_USER_TEXT, show_alert=True)
+            return ConversationHandler.END
```
Т.к. здесь ты уже получаешь профиль, то можно сразу проверить, что пользователь забанен без check_user_ban, где совершается ещё один запрос на получение профиля. 
```Python
if context.user_data["profile_info"].is_banned:
```

TomatoInOil commented on src/bot/conversations/roommate_search/callback_funcs.py:
```diff
@@ -26,6 +28,11 @@ async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> int:
+    banned_user = await check_user_ban(update)
+    if banned_user:
+        await update.callback_query.answer(text=BANNED_USER_TEXT, show_alert=True)
+        return ConversationHandler.END
```
В функциях `start()` из 
- `src/bot/conversations/roommate_search/callback_funcs.py`
- `src/bot/conversations/coliving_search/callback_funcs.py`
- `src/bot/conversations/coliving/callback_funcs.py`

Есть 2 общих момента: 
1. Эта логика проверки, что пользователь не забанен 
2. Декоратор `@profile_required`

Можно перенести эту логику в декоратор `profile_required`, чтобы не совершать ещё один запрос к API на получение профиля

Есть ещё один случай использования `@profile_required`
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/386529702ccb376046797914b7c7ebb347df294b/src/bot/conversations/invitation/callback_funcs.py#L14-L20
Но я думаю, что проверка, что пользователь не в бане не помешает здесь

TomatoInOil commented on src/backend/profiles/admin.py:
```diff
@@ -10,15 +10,29 @@ class UserFromTelegramAdmin(admin.ModelAdmin):
         "telegram_id",
+        "user_profile",
         "residence",
     )
```
Сюда можно ещё добавить Inline отображение для коливингов пользователя, чтобы админу не искать их. 
https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#inlinemodeladmin-objects

NRenat replied:
Вместо Inline реализовал ссылку на коливинг.

TomatoInOil commented on src/backend/profiles/admin.py:
```diff
@@ -10,15 +10,29 @@ class UserFromTelegramAdmin(admin.ModelAdmin):
 
+    def save_model(self, request, obj, form, change):
+        coliving = Coliving.objects.filter(host=obj).first()
+        if obj.is_banned:
+            obj.user_profile.is_visible = False
+            obj.user_profile.save()
+
+            if coliving:
+                coliving.is_visible = False
+                coliving.save()
```
Чтобы у других пользователей не было проблем со взаимодействием с этим пользователем, давай удалим лайки, связанные с ним
```Python
ProfileLike.objects.filter(Q(sender=profile) | Q(receiver=profile))
ColivingLike.objects.filter(Q(sender=profile) | Q(coliving=coliving))
```
, а также установим `obj.residence` в `None`

TomatoInOil reviewed (changes requested):
Заблокированному пользователю не имеет смысла отправлять рассылку, поэтому можно сюда добавить filter
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/b4641d22f834c9bdbfa1a9e43508f502f820f914/src/backend/mailing/api_views.py#L34

NRenat committed:
review
- В админку добавлена ссылка на коливинг пользователя, если он им владеет.
- Добавлено выписывание заблокированных пользователей из коливингов.
- Добавлено удаление лайков заблокированных пользователей.
- Добавлена обработка случаев получения лайков заблокированных пользователей.
- Перенесена проверка на заблокированного пользователя в декоратор @profile_required.

NRenat committed:
resolve conflicts 0006_remove_mailing_is_sent_mailing_status_and_more.py

NRenat committed:
Добавлен знак "№" для лучшей читаемости.
